src/appleDeviceUtils/appleDeviceUtils.js

```diff
@@ -0,0 +1,105 @@
+// Do not remove -- Apple devices change this when they support modern webkit messaging
+let hasModernWebkitAPI = false
+// INJECT hasModernWebkitAPI HERE
+
+// The native layer will inject a randomised secret here and use it to verify the origin
+const secret = 'PLACEHOLDER_SECRET'
+
+const ddgGlobals = window.navigator.ddgGlobals
+
+/**
+ * Sends message to the webkit layer
+ * @param {String} handler
+ * @param {*} data
+ * @returns {*}
+ */
+const wkSend = (handler, data = {}) =>
+    window.webkit.messageHandlers[handler].postMessage(data)
+
+/**
+ *
+ * @param {String} randomMethodName
+ * @param {Function} callback
+ */
+const generateRandomMethod = (randomMethodName, callback) => {
+    Object.defineProperty(ddgGlobals.window, randomMethodName, {
+        enumerable: false,
+        // configurable, To allow for deletion later
+        configurable: true,
+        writable: false,
+        // Use proxy to ensure stringification isn't possible
+        value: new Proxy(function () {}, {
+            apply (target, thisArg, args) {
+                callback(...args)
+                delete ddgGlobals.window[randomMethodName]
+            }
+        })
+    })
+}
+
+/**
+ * Sends message to the webkit layer and waits for the specified response
+ * @param {String} handler
+ * @param {*} data
+ * @returns {Promise<*>}
+ */
+const wkSendAndWait = async (handler, data = {}) => {
+    if (hasModernWebkitAPI) {
+        return wkSend(handler, data)
+    }
+
+    // Older versions
+    const randMethodName = createRandMethodName()
+    const key = await createRandKey()
+    const iv = createRandIv()
+
+    const encryptedResponse = await new Promise((resolve) => {
```

Comment:
Should we also save `Promise` in our globals, given that we're using it to return the data? Given that at this stage it's still encrypted, we shouldn't need it here, but I wonder if there is a way to compromise it that could expose unencrypted data. This whole `wkSendAndWait` is `async`, after all. Not sure 🤔.

Reply:
Yeah a hooked Promise implementation could call it's own code first something like:
```
Promise = class Promise {
  constructor(callback) {
    const res = (data) =>{xhr(..., data)}
    callback(res)
  }
}
```

See: https://github.com/taylorhakes/promise-polyfill/blob/master/src/index.js is a good example of a shim here.

Reply:
Saving `Promise` itself is sufficient, right?

Reply:
I'm now using `ddgGlobals.Promise`.